File: benchmarks/mandelbrot/bench.py
import time
W, H, M = 400, 300, 256
def mandelbrot(w, h, max_iter):
    total = 0
    for py in range(h):
        for px in range(w):
            cx = -2.5 + px / w * 3.5
            cy = -1.2 + py / h * 2.4
            zx = zy = 0.0; it = 0
            while it < max_iter:
                zx2, zy2 = zx*zx, zy*zy
                if zx2 + zy2 > 4.0: break
                zx, zy = zx2 - zy2 + cx, 2*zx*zy + cy; it += 1
            total += it
    return total
t = time.perf_counter()
total = mandelbrot(W, H, M)
print(f"Mandelbrot {W}x{H} maxIter={M}: total_iters={total} ({(time.perf_counter()-t)*1000:.1f}ms)")

File: benchmarks/mandelbrot/bench.rs
use std::time::Instant;
fn mandelbrot(width: u32, height: u32, max_iter: u32) -> u64 {
    let mut total: u64 = 0;
    for py in 0..height {
        for px in 0..width {
            let cx = -2.5 + (px as f64) / (width as f64) * 3.5;
            let cy = -1.2 + (py as f64) / (height as f64) * 2.4;
            let (mut zx, mut zy) = (0.0f64, 0.0f64);
            let mut iter = 0u32;
            while iter < max_iter {
                let zx2 = zx * zx;
                let zy2 = zy * zy;
                if zx2 + zy2 > 4.0 { break; }
                let tmp = zx2 - zy2 + cx;
                zy = 2.0 * zx * zy + cy;
                zx = tmp;
                iter += 1;
            }
            total += iter as u64;
        }
    }
    total
}
fn main() {
    let (w, h, m) = (400u32, 300u32, 256u32);
    let t = Instant::now();
    let total = mandelbrot(w, h, m);
    println!("Mandelbrot {}x{} maxIter={}: total_iters={} ({}ms)", w, h, m, total, t.elapsed().as_millis());
}
